keps/sig-api-machinery/20200309-consistent-resource-versions-semantics.md

```diff
@@ -0,0 +1,216 @@
+status: provisional
+---
+
+# Title
+
+consistent-resource-version-semantics
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+- [Proposal](#proposal)
+  - [Add a ResourceVersionMatch query parameter](#add-a-resourceversionmatch-query-parameter)
+  - [Backward Compatibility](#backward-compatibility)
+  - [Get support?](#get-support)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Alternatives Considered](#alternatives-considered)
+  - [Alternative: Introduce ExactResourceVersion and MinResourceVersion parameters](#alternative-introduce-exactresourceversion-and-minresourceversion-parameters)
+  - [Alternative: Use syntax in the query string](#alternative-use-syntax-in-the-query-string)
+<!-- /toc -->
+
+## Summary
+
+Make resource version semantics consistent for list and get requests regardless of
+pagination.
+
+## Motivation
+
+Resource version semantics are inconsistent when using pagination. When a list
+request is made with a resourceVersion but no limit, “Not older than” semantics
+apply, but once a limit is set, “Exact” semantics apply. See [API Concepts: Resource Version](https://kubernetes.io/docs/reference/using-api/api-concepts/#resource-versions)
+for details about the semantics.
+
+The inconsistency is confusing and problematic. A client expecting “Not older
+than” semantics has no reason to expect the “410 Gone” HTTP responses from the
+server that can be sent when “Exact” semantics are applied and when the
+requested resource version is older than what is available in watch cache or the
+etcd compaction window. Even if the author of code to make a list request was
+aware of this inconsistency, the code to make a list request might be separated
+from the code setting a limit on the request. For example, ListerWatchers can be
+layered in to add pagination limits.
+
+This was discussed in [Feb 12th, 2020 SIG API machinery bi-weekly meeting](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#bookmark=id.3kvpricxohe8).
+
+### Goals
+
+- Improve API to make resource version semantics consistent.
+- Backward compaibility. In particular, support the case of a client making a request using the
+  updated API against an server serving the old API.
+
+## Proposal
+
+### Add a ResourceVersionMatch query parameter
+
+Add an optional `ResourceVersionMatch` paramater to `ListOptions` and
+`GetOptions` with the enumeration values:
+
+```
+// ResourceVersionMatch specifies how the ResourceVersion parameter is applied. ResourceVersionMatch
+// may only be set if ResourceVersion is also set.
+type ResourceVersionMatch string
+
+const (
+	// ResourceVersionMatchNotOlderThan matches data at least as new as the provided
+	// ResourceVersion. The newest available data is preferred, but any data not
+	// older than this ResourceVersion may be served.
+	// This guarantees that ResourceVersion in the ListMeta is not older than the requested
```

Comment:
nit: this part of comment is very focused on LIST operation - it doesn't make sense neither for GET not for watch...